### PR TITLE
fix #149 retry ConnectionError on fetch_range

### LINKED ISSUE #149: occasionally getting ConnectionResetError while reading large files (100MB+)
Hi you've got great lib (just switched from `smart_open` for SSE support) so thanks!

at my work we're using S3 as main storage for ETL and using this lib currently for all io (multi million file reads per day).
we're getting the following error rarely (several times a day) almost all of them occurs when reading files larger than 100MB.

the files are gzipped and some are tarred but it seems the only characteristic for this error is file size.

```
# trimmed stacktrace
File "/MYPROJECT/MY_FUNC.py", line 123, in <genexpr>
rows = (row for row in reader)
File "/XXX/lib/python3.6/csv.py", line 112, in _next_
row = next(self.reader)
File "/XXX/lib/python3.6/tarfile.py", line 706, in readinto
buf = self.read(len(b))
File "/XXX/lib/python3.6/tarfile.py", line 695, in read
b = self.fileobj.read(length)
File "/XXX/lib/python3.6/gzip.py", line 276, in read
return self._buffer.read(size)
File "/XXX/lib/python3.6/_compression.py", line 68, in readinto
data = self.read(len(byte_view))
File "/XXX/lib/python3.6/gzip.py", line 469, in read
buf = self._fp.read(io.DEFAULT_BUFFER_SIZE)
File "/XXX/lib/python3.6/gzip.py", line 91, in read
self.file.read(size-self._length+read)
File "/XXX/lib/python3.6/site-packages/s3fs/core.py", line 1311, in read
self._fetch(self.loc, self.loc + length)
File "/XXX/lib/python3.6/site-packages/s3fs/core.py", line 1292, in _fetch
req_kw=self.s3.req_kw)
File "/XXX/lib/python3.6/site-packages/s3fs/core.py", line 1496, in _fetch_range
return resp['Body'].read()
File "/XXX/lib/python3.6/site-packages/botocore/response.py", line 74, in read
chunk = self._raw_stream.read(amt)
File "/XXX/lib/python3.6/site-packages/botocore/vendored/requests/packages/urllib3/response.py", line 239, in read
data = self._fp.read()
File "/XXX/lib/python3.6/http/client.py", line 462, in read
s = self._safe_read(self.length)
File "/XXX/lib/python3.6/http/client.py", line 612, in _safe_read
chunk = self.fp.read(min(amt, MAXAMOUNT))
File "/XXX/lib/python3.6/socket.py", line 586, in readinto
return self._sock.recv_into(b)
File "/XXX/lib/python3.6/ssl.py", line 1009, in recv_into
return self.read(nbytes, buffer)
File "/XXX/lib/python3.6/ssl.py", line 871, in read
return self._sslobj.read(len, buffer)
File "/XXX/lib/python3.6/ssl.py", line 631, in read
v = self._sslobj.read(len, buffer)
ConnectionResetError: [Errno 104] Connection reset by peer
```

its tricky to retry a file reading when we read it as stream at our code base currently so we solve this using patching `S3_RETRYABLE_ERRORS`.

digging a bit inside s3fs sources I found:

```python
core.S3_RETRYABLE_ERRORS = (socket.timeout, )
```

so now we are patching s3fs dynamically to solve this as follow:

```python
s3fs.core.S3_RETRYABLE_ERRORS = s3fs.core.S3_RETRYABLE_ERRORS + (OSError, )
```

I am willing to submit a patch to solve this issue at s3fs level - add the relevant exceptions to `S3_RETRYABLE_ERRORS` or solve this in any other way.

Please LMK if you're interested in fixing this issue.

Thanks!
Shmulik

### PR DESCRIPTION
**Pull request by shmuelamar**
1. I didnt see any test for the other retry logic so i assume its ok, LMK if this needs a test and what approach to use to simulate the exception
2. to support both py2 and py3 i had to write it like this (as the comments explain). LMK if you want this only on py3 or its ok.

thanks,
Shmulik

### CODE REVIEW

**Comment by shmuelamar** [comment]
is the failing build related to my change? shall I try address it in this PR?

**Comment by martindurant** [comment]
> is the failing build related to my change?

I don't immediately see why your change would have caused the failure, but it needs to be investigated.

File: s3fs/core.py [hunk 1]
```diff
@@ -1498,6 +1509,14 @@ def _fetch_range(client, bucket, key, version_id, start, end, max_attempts=10,
             logger.debug('Exception %e on S3 download, retrying', e,
                          exc_info=True)
             continue
+        except OSError as e:
+            # only retry connection_errors - similar to catching
+            # ConnectionError on py3
+            if e.errno not in _CONNECTION_ERRORS:
+                raise
+            logger.debug('ConnectionError %e on S3 download, retrying', e,
+                         exc_info=True)
```
[the remark this answers is not in the document]
Comment (martindurant) [reply s3fs/core.py]: OK, I suppose you're right, if it's rare, full information is reasonable.

**Comment by martindurant** [comment]
I seem to be seeing similar errors locally after updating boot3/botocore

**Comment by shmuelamar** [comment]
> I seem to be seeing similar errors locally after updating boot3/botocore

thanks for the quick replies @martindurant :). LMK if there is something i can help fixing this.

**Comment by martindurant** [comment]
https://github.com/spulec/moto/issues/1793

The solution appears to be to specify the older botocore version in the test_requirememnts.txt . Can you try this?

**Comment by shmuelamar** [comment]
> [spulec/moto#1793](https://github.com/spulec/moto/issues/1793)
> 
> The solution appears to be to specify the older botocore version in the test_requirememnts.txt . Can you try this?

ok now build passes :smiley:

**Comment by martindurant** [comment]
Thank you.

**Pull request merged**